torch/distributed/_composable/fsdp/_fsdp_collectives.py

```diff
@@ -15,7 +15,13 @@ class AllGatherResult(NamedTuple):
     all_gather_output: torch.Tensor
     all_gather_event: Optional[torch.cuda.Event]
     all_gather_work: Optional[dist.distributed_c10d.Work]
-    all_gather_input_numels: List[int]
+    # For each parameter, the all-gather input dtype for each input
+    param_all_gather_input_dtypes: List[List[torch.dtype]]
```

Comment:
would there be multiple input dtypes desired for each param? IMO if there's a single dtype then we don't need this field.

Reply:
Yes, we need this to support different dtypes (like mixing fp8 and bf16).

Reply:
One logical parameter could contribute >1 inner tensors with different dtypes (e.g. data in some fp8 dtype and scales in fp32).